Opus: Avoid vector crashloops

paweljw opened the pull request:
I managed to reproduce the bug we’ve been seeing, though only intermittently. It occurs under the most (not necessarily all) following conditions:

- Vector is under heavy memory pressure -> disk buffers are in use
- File system watchers/inotify are near exhaustion (many log files watched/files open on the system)
- An HTTP sink fails or applies backpressure, increasing memory/disk buffer pressure
- A SIGHUP arrives, causing Vector to try and re-establish watches, exhausting open files limit
- Can also be exacerbated by an oomkill on Vector (memory pressure from buffers plus, I assume, file watcher metadata)

When this happens, Vector fails to reload its configuration. The process fails, then Supervisor restarts it, treating the exit codes as normal. On restart, Vector can’t access its configuration files, so it comes up without the proper settings or buffer configuration. Once it enters this state, it stays there in fail-safe mode until something intervenes. This last part is likely a Vector bug - but I don't have concrete enough reproduction steps to feel good about reporting it.

It seems reasonable this would most readily occur in Kubernetes:
- enforced memory limits
- likely many files in /host/var/log for Vector to watch
- kubernetes logs source watching pod files
- memory pressure on the node -> pod restarts -> more pod logs to watch -> more memory pressure from Vector

It also explains symptops well:
- no logs sunk -> Vector is running with failsafe config (no proper sinks)
- no buffer files -> Vector is running with failsafe config (no buffers)
- logs show Vector running -> yes, just not sinking anything into nonexistent http sinks

To address this, I introduced the following changes:
1.	Config validation at startup – if /vector-config/current is inaccessible and no previous valid config can be restored, the startup script fails immediately. If open files limit is exhausted, the check script should be unable to access configuration as well.
2.	Health check script – detects when Vector is stuck in fail-safe mode (only the default console sink present). If so, it forces a restart, counting against Supervisor’s restart limit (now 10).
3.	Exit code handling – exit codes 0,1,2 are considered normal and restartable; higher codes are treated as fatal, abandoning the failing pod instead of retrying endlessly.

This setup should help automatically recover from cases where Vector is “limping” along in fail-safe mode and pretending to be healthy. In its harshest form, this should take the form of abandoning the pod (and the potentially unhealthy overlay filesystem), and starting fresh. Ditto for docker; we likely haven't seen it in Docker installations because it's just harder to churn/impact a Docker Compose/Swarm service enough.

This may just be the product of time + churn - Vector will exhaust open files eventually when tightly limited. The timing is suspicious though: we only see it now, after I rolled out Helm/Compose changes mounting the host filesystem read-only under /host. In theory, exposing more files shouldn’t matter (exposed ≠ watched), but that correlation led me to simulate extreme logfile churn and resource constraints, which reproduced the bug. If such issues persist, I would consider rolling back that change and reassessing how we approach custom log files.

My intent here is to both reduce the immediate triggers (watcher/file exhaustion) and add safeguards so Vector can’t silently remain non-functional in the future.

This also makes Vector API non-optional (necessary for health checks). I will propose a patch in Telemetry that makes it the default.

paweljw commented:
@gyfis happy to change direction here, though I feel relatively strongly the root approach at least makes sense 🙏

curusarn commented on CLAUDE.md:
```diff
@@ -146,10 +146,57 @@ The container always exposes:
+- Attempts to restore from `/vector-config/latest-valid-upstream/` if current is missing
+- Exits with code 127 (critical failure) if no configs found, triggering supervisor retry
+
+#### 2. Atomic Configuration Updates (`engine/vector_config.rb`)
+- Uses symlinks for `/vector-config/current` instead of moving directories
+- Creates temp symlink first, then atomically renames to 'current'
+- Backs up previous config to 'previous' link before switching
+- Restores backup if promotion fails, preventing partial states
```
After Re-reading the code: It makes sense and now the promotion is actually atomic. 

I don't love the extra complexity. Especially filtering out "in use" directories during clean up. But I don't have a better solution in mind.

paweljw commented:
@curusarn 

* removed the sysctls - default limits should be fine
* moved the healthchecking to a docker compose healthcheck (which also applies to Swarm)

curusarn commented on healthcheck.sh:
```diff
@@ -0,0 +1,45 @@
+SINK_COUNT=$(echo "$SINKS" | wc -l)
+
+# Check if only console sink exists (emergency/fallback mode)
+if [ "$SINK_COUNT" -eq 1 ] && echo "$SINKS" | grep -q "^console$"; then
+    echo "ERROR: Vector running with console-only sink (lost configuration)"
+    exit 1
+fi
```
Is this going to crash loop collector if I use manual config with just one sink? Could we check for specific sink name?

paweljw replied:
Even if you only have one sink in manual, the generated config has 3 sinks. Also, this line checks if that 1 sink is the default `console` sink. Assuming the generated config was picked up, this should never be 1.

curusarn replied:
I was under impression that if you uncheck all checkboxes you get empty generated config.

paweljw replied:
No, I just checked (unchecked everything in a Kubernetes collector) - we always build the Better Stack http sinks, with the catch-all inputs for manual configs.

paweljw replied:
Which makes sense I think - even in a manual config, we show how to sink data into the provided sinks, so the manually-configured data shows up in Better Stack :)

curusarn reviewed (approved):
Thank you! Looks good.

I would make the health check more specific to the default configuration we expect in cases when upstream configs are lost. 
We don't want to keep restarting for custom manual configs.
